akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala

```diff
@@ -514,46 +617,43 @@ private[akka] class Shard(
   private def receiveTerminated(ref: ActorRef): Unit = {
     if (handOffStopper.contains(ref))
       context.stop(self)
-    else if (idByRef.contains(ref) && handOffStopper.isEmpty)
-      entityTerminated(ref)
   }
 
   @InternalStableApi
-  def entityTerminated(ref: ActorRef): Unit = {
+  def entityTerminated(ref: ActorRef, id: EntityId): Unit = {
     import settings.tuningParameters._
-    val id = idByRef(ref)
-    idByRef -= ref
-    refById -= id
+    val isPassivating = entities.isPassivating(id)
+    entities.terminated(ref)
     if (passivateIdleTask.isDefined) {
       lastMessageTimestamp -= id
     }
     if (messageBuffers.getOrEmpty(id).nonEmpty) {
       // Note; because we're not persisting the EntityStopped, we don't need
       // to persist the EntityStarted either.
       log.debug("Starting entity [{}] again, there are buffered messages for it", id)
+      // this will re-add the entity back
       sendMsgBuffer(id)
     } else {
-      if (!passivating.contains(ref)) {
+      if (!isPassivating) {
         log.debug("Entity [{}] stopped without passivating, will restart after backoff", id)
-        // note that it's not removed from state here, will be started again via RestartEntity
+        entities.waitingForRestart(id)
         import context.dispatcher
         context.system.scheduler.scheduleOnce(entityRestartBackoff, self, RestartEntity(id))
       } else {
         // FIXME optional wait for completion as optimization where stops are not critical
-        waitForAsyncWrite(id, RememberEntitiesShardStore.RemoveEntity(id))(_ => passivateCompleted(id))
+        sendToRememberStore(id, RememberEntitiesShardStore.RemoveEntity(id))(_ => passivateCompleted(id))
```

Comment:
mentioned in the other PR review, but perhaps either don't take a id param to the whenDone or actually use it.

Reply:
Yep i'll do that here